### PR TITLE
callback email store in a dedicated folder and check number submitted status email

### PR DESCRIPTION
**Pull request by burnout87**

### CODE REVIEW

**Comment by burnout87** [comment]
- new parameters in the configuration (added to the tests fixture and also the example)
- relative new token field
- adapted tests

File: cdci_data_analysis/config_dir/conf_env.yml.example [hunk 1]
```diff
@@ -48,6 +48,9 @@ dispatcher:
         email_sending_timeout_default_threshold: 1800
         # sending of an email when a request has been submitted
         email_sending_job_submitted: True
+        # amount of time (in seconds) elapsed from the sending of the last email
+        # for when a request has been submitted
+        email_sending_job_submitted_default_interval: True
```
Comment (volodymyrss) [inline cdci_data_analysis/config_dir/conf_env.yml.example:53]: True in seconds?

File: cdci_data_analysis/flask_app/dispatcher_query.py [hunk 2]
```diff
@@ -595,6 +599,25 @@
...
+        # to be handled now, with the job_id generated taking into account only the user_id
+        query_output_json_file_aliased = f'scratch_sid_{session_id}_jid_{job_id}_aliased'
+        if os.path.exists(query_output_json_file_aliased):
+            path_email_history_folder = query_output_json_file_aliased + '/email_history'
+        elif os.path.exists(query_output_json_file):
+            path_email_history_folder = query_output_json_file + '/email_history'
+        if not os.path.exists(path_email_history_folder):
+            os.makedirs(path_email_history_folder)
+        email_files_list = glob.glob(path_email_history_folder + '/email_*')
+        number_emails_scartch_dirs = len(email_files_list)
```
Comment (volodymyrss) [inline cdci_data_analysis/flask_app/dispatcher_query.py:614]: scartch => scratch

File: tests/test_job_management.py [hunk 3]
```diff
@@ -160,7 +168,13 @@ def test_email_callback_after_run_analysis(dispatcher_live_fixture, dispatcher_l
         assert 'email_status' not in jdata
     else:
         assert jdata['exit_status']['email_status'] == 'email sent'
-        smtp_server_log = f'local_smtp_log/{dispatcher_local_mail_server.id}_local_smtp_output.json'
+
+        # check the email in the email folders, and that the first one was produced
+        assert os.path.exists(email_history_folder_path)
+        assert os.path.exists(email_history_folder_path)
```
Comment (volodymyrss) [inline tests/test_job_management.py:174]: twice the line?

File: tests/test_job_management.py [hunk 4]
```diff
@@ -160,7 +168,13 @@ def test_email_callback_after_run_analysis(dispatcher_live_fixture, dispatcher_l
...
+        list_email_files = glob.glob(email_history_folder_path + '/email_0_submitted_*.email')
+        assert len(list_email_files) == 1
```
Comment (volodymyrss) [inline tests/test_job_management.py:176]: can we simulate the case when one mail was discarded?

File: cdci_data_analysis/config_dir/conf_env.yml.example (hunk 1 above)
Comment (burnout87) [reply cdci_data_analysis/config_dir/conf_env.yml.example:53]: damn, my bad, will fix

File: tests/test_job_management.py (hunk 4 above)
Comment (burnout87) [reply tests/test_job_management.py:176]: will do in the next commit

File: cdci_data_analysis/flask_app/dispatcher_query.py [hunk 5]
```diff
@@ -611,18 +634,18 @@ def generate_products_url_from_par_dict(self, products_url, par_dict) -> str:
...
         request_url = ""
         file = None
         if self.scratch_dir:
             file = open(self.scratch_dir + '/query_output.json')
         else:
-            if os.path.exists(job_monitor_status_json_file_aliased):
-                file = open(job_monitor_status_json_file_aliased)
-            elif os.path.exists(job_monitor_status_json_file):
-                file = open(job_monitor_status_json_file)
+            if os.path.exists(query_output_json_file_aliased):
```
Comment (volodymyrss) [inline cdci_data_analysis/flask_app/dispatcher_query.py:645]: so, 645-648 not covered by test, codecov complains? scratch existed?

File: cdci_data_analysis/pytest_fixtures.py [hunk 6]
```diff
@@ -288,6 +288,7 @@ def dispatcher_test_conf_fn(tmpdir):
         email_sending_timeout: True
         email_sending_timeout_default_threshold: 1800
         email_sending_job_submitted: True
+        email_sending_job_submitted_default_interval: 5
```
Comment (volodymyrss) [inline cdci_data_analysis/pytest_fixtures.py:291]: I would make something like 60, but it's minor and irrelevant

**Review by volodymyrss — APPROVED** [review]
Small comments, as you wish.

**Comment by volodymyrss** [comment]
found another case: I requested publicly, then wanted to resubmit with account. it started with `progress` - no emails got.
is it normal? is scratch directory not custom per user?
in any case, I expected an email

File: cdci_data_analysis/flask_app/dispatcher_query.py (hunk 5 above)
Comment (burnout87) [reply cdci_data_analysis/flask_app/dispatcher_query.py:645]: probably this check is not really necessary

Comment (burnout87) [reply cdci_data_analysis/flask_app/dispatcher_query.py:645]: the scratch directory is always set

**Comment by burnout87** [comment]
> found another case: I requested publicly, then wanted to resubmit with account. it started with `progress` - no emails got.
> is it normal? is scratch directory not custom per user?
> in any case, I expected an email

Did you write a test for this? Or was it done from the frontend? So that I can reproduce.

Also, Which email configuration did you use?

**Comment by volodymyrss** [comment]
> > found another case: I requested publicly, then wanted to resubmit with account. it started with `progress` - no emails got.
> > is it normal? is scratch directory not custom per user?
> > in any case, I expected an email
> 
> Did you write a test for this? Or was it done from the frontend? So that I can reproduce.
> 
> Also, Which email configuration did you use?

With frontend. Test is possible but I did not do it.
Is the flow clear? 
There could be other unusual flows which may cause unexpected behavior. I make a separate issue for it.

**Comment by burnout87** [comment]
> > > found another case: I requested publicly, then wanted to resubmit with account. it started with `progress` - no emails got.
> > > is it normal? is scratch directory not custom per user?
> > > in any case, I expected an email
> > 
> > 
> > Did you write a test for this? Or was it done from the frontend? So that I can reproduce.
> > Also, Which email configuration did you use?
> 
> With frontend. Test is possible but I did not do it.
> Is the flow clear?
> There could be other unusual flows which may cause unexpected behavior. I make a separate issue for it.

Yes it' clear, seems a bit tricky to reproduce, but it has to be traced.

**Comment by burnout87** [comment]
I guess this can be merged. Then we can continue with more specific tests

**Review by volodymyrss — APPROVED** [review]

**Pull request merged**